File: java/ReverseInteger.java
/*
 * Filename: ReverseInteger.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Wed 04 Mar 2015 06:04:38 PM EST
 */

public class Solution {
	public static int reverse(int x) {
		int digit;
		int rest = x;
		int res = 0;

		while (rest != 0) {
			digit = rest % 10;
			rest = rest / 10;

			if (x > 0) {
				if (res > Integer.MAX_VALUE / 10 ||
						(res == Integer.MAX_VALUE / 10 && digit > Integer.MAX_VALUE % 10)) {
					return 0;
				} else {
					res = res * 10 + digit;
				}
			} else if (x < 0) {
				if (res < Integer.MIN_VALUE / 10 ||
					(res == Integer.MIN_VALUE / 10 && digit < Integer.MIN_VALUE % 10)) {
						return 0;
				} else {
					res = res * 10 + digit;
				}
			}
		}

		return res;
	}
}

File: cpp/reverseInteger.cpp
/*
 * Filename: reverseInteger.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Thu 21 Jan 2016 08:35:09 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

class Solution
{
	public:
		int reverse(int x)
		{
			int digit;
			int res = 0;

			while (x)
			{
				digit = x % 10;

				if (x > 0 && (res > INT_MAX / 10 || (res == INT_MAX / 10 && digit > INT_MAX % 10)) ||
						x < 0 && (res < INT_MIN / 10 || (res == INT_MIN / 10 && digit < INT_MIN % 10)))
				{
					return 0;
				}

				x /= 10;
				res = res * 10 + digit;
			}

			return res;
		}
};
